File: packages/agent-middlewares/src/filesystem.ts
/**
 * Filesystem Middleware
 *
 * Encapsulates all filesystem tools for file and directory operations.
 * Provides tools for reading, writing, searching, and managing files.
 *
 * ## Available Tools
 *
 * - **read_file**: Read file contents with optional line limits
 * - **write_file**: Write content to files
 * - **edit_file**: Perform exact string replacements
 * - **glob_files**: Find files by name patterns
 * - **search_files_rg**: Fast text search using ripgrep
 * - **folder_operations**: Create, list, and check folder existence
 *
 * ## Usage
 *
 * ```typescript
 * import { FilesystemMiddleware } from '@langgraph-js/agent-middlewares';
 *
 * const fsMiddleware = new FilesystemMiddleware();
 *
 * const agent = createAgent({
 *   model,
 *   systemPrompt,
 *   tools,
 *   middleware: [fsMiddleware]
 * });
 * ```
 */

import { AgentMiddleware } from 'langchain';
import {
    read_tool,
    write_tool,
    replace_tool,
    glob_tool,
    grep_tool,
    folder_tool,
} from './tools/filesystem_tools/index.js';

/**
 * Middleware that provides filesystem operations capabilities.
 *
 * This middleware encapsulates all filesystem-related tools:
 * - File I/O: read, write, edit
 * - File search: glob, grep
 * - Folder operations: create, list, exists
 *
 * All paths are resolved relative to the current working directory (cwd)
 * stored in the agent state, providing flexible path handling.
 */
export class FilesystemMiddleware implements AgentMiddleware {
    name = 'FilesystemMiddleware';

    // No additional context schema needed
    contextSchema = undefined;

    // No state schema modification needed
    stateSchema = undefined;

    // Export all filesystem tools
    tools = [read_tool, write_tool, replace_tool, glob_tool, grep_tool, folder_tool];

    /**
     * Execute method for AgentPackage compatibility
     * Returns self to support initialization without parameters
     */
    async execute(): Promise<this> {
        return this;
    }
}

export type { BaseAgentStateType } from './index.js';

File: rust-agent-middlewares/src/middleware/filesystem.rs
use async_trait::async_trait;
use rust_create_agent::agent::state::State;
use rust_create_agent::middleware::r#trait::Middleware;
use rust_create_agent::tools::{BaseTool, ToolProvider};

use crate::tools::{
    EditFileTool, FolderOperationsTool, GlobFilesTool, ReadFileTool, SearchFilesRgTool,
    WriteFileTool,
};

/// FilesystemMiddleware - 与 TypeScript FilesystemMiddleware 对齐
pub struct FilesystemMiddleware;

impl FilesystemMiddleware {
    pub fn new() -> Self {
        Self
    }

    pub fn build_tools(cwd: &str) -> Vec<Box<dyn BaseTool>> {
        vec![
            Box::new(ReadFileTool::new(cwd)),
            Box::new(WriteFileTool::new(cwd)),
            Box::new(EditFileTool::new(cwd)),
            Box::new(GlobFilesTool::new(cwd)),
            Box::new(SearchFilesRgTool::new(cwd)),
            Box::new(FolderOperationsTool::new(cwd)),
        ]
    }

    pub fn tool_names() -> Vec<&'static str> {
        vec![
            "read_file",
            "write_file",
            "edit_file",
            "glob_files",
            "search_files_rg",
            "folder_operations",
        ]
    }
}

impl Default for FilesystemMiddleware {
    fn default() -> Self {
        Self::new()
    }
}

impl ToolProvider for FilesystemMiddleware {
    fn tools(&self, cwd: &str) -> Vec<Box<dyn BaseTool>> {
        Self::build_tools(cwd)
    }
}

#[async_trait]
impl<S: State> Middleware<S> for FilesystemMiddleware {
    fn collect_tools(&self, cwd: &str) -> Vec<Box<dyn BaseTool>> {
        Self::build_tools(cwd)
    }

    fn name(&self) -> &str {
        "FilesystemMiddleware"
    }
}
